src/VisualStudio/Core/Def/Implementation/ProjectSystem/DocumentProvider.cs

```diff
@@ -175,22 +206,115 @@ private ITextBuffer TryGetTextBufferFromDocData(object docData)
+                AddCookieOpenDocumentPair_NoLock(docCookie, documentKey);
             }
         }
 
         /// <summary>
         /// Notifies the document provider that this document is now registered in a project.
+        /// If we are on a foregroud thread, then this is done right away.
+        /// Otherwise, we schedule a task on foreground task scheduler.
         /// </summary>
         public void NotifyDocumentRegisteredToProject(IVisualStudioHostDocument document)
         {
+            if (IsForeground())
+            {
+                NotifyDocumentRegisteredToProject_Core(document, cancellationToken: CancellationToken.None);
+            }
+            else
+            {
+                var cts = new CancellationTokenSource();
+                var task = InvokeBelowInputPriority(() => NotifyDocumentRegisteredToProject_Core(document, cts.Token), cts.Token);
+                AddPendingDocumentInitializationTask(document, task, cts);
+            }
+        }
+
+        private void AddPendingDocumentInitializationTask(IVisualStudioHostDocument document, Task task, CancellationTokenSource cts)
+        {
+            var taskAndTokenSource = new TaskAndTokenSource() { Task = task, CancellationTokenSource = cts };
+            lock (_gate)
+            {
+                // Add taskAndTokenSource to the pending document initialization tasks.
+                // Check for cancellation before adding as the task might already have been completed/cancelled/faulted before we reached here.
+                if (!cts.IsCancellationRequested && !task.IsCompleted && !task.IsFaulted)
+                {
+                    _pendingDocumentInitializationTasks.Add(document.Id, taskAndTokenSource);
+                }
+            }
+        }
+
+        private void CancelPendingDocumentInitializationTask(IVisualStudioHostDocument document)
+        {
+            lock (_gate)
+            {
+                CancelPendingDocumentInitializationTask_NoLock(document);
+            }
+        }
+
+        private void CancelPendingDocumentInitializationTasks_NoLock(IEnumerable<DocumentKey> documentKeys)
+        {
+            foreach (var documentKey in documentKeys)
+            {
+                StandardTextDocument document;
+                if (_documentMap.TryGetValue(documentKey, out document))
+                {
+                    CancelPendingDocumentInitializationTask_NoLock(document);
+                }
+            }
+        }
+
+        private void CancelPendingDocumentInitializationTask_NoLock(IVisualStudioHostDocument document)
+        {
+            // Cancel pending initialization task for the document, if any, and dispose the cancellation token source.
+            TaskAndTokenSource taskAndTokenSource;
+            if (_pendingDocumentInitializationTasks.TryGetValue(document.Id, out taskAndTokenSource))
+            {
+                // Signal cancellation and wait for task to complete.
```

Comment:
I wasn't sure if running task might access data structures with the assumption of the document still being active, while the callee doing the cancellation might proceed after this invocation and remove the document entry from the data structures.